midend/replaceSelectRange.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+* Copyright 2020, MNK Labs & Consulting
+* http://mnkcg.com
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+
+#include "replaceSelectRange.h"
+
+namespace P4 {
+
+static int trailingZeros(big_int n, int width) {
+    int zeros = 0;
+
+    while (n > 0 && (n & 1) == 0) {
+        zeros += 1;
+        n >>= 1;
+    }
+
+    return (zeros < width) ? zeros : width;
+}
+
+bool DoReplaceSelectRange::checkRange(const IR::Range* range) {
+    if (range == nullptr)
+        return false;
+
+    if (!range->left->is<IR::Constant>()) {
+        ::error("%1%: must evaluate to a compile-time constant.",
+                range->left);
+        return false;
+    }
+    auto left = range->left->to<IR::Constant>()->value;
+    auto base = range->left->to<IR::Constant>()->base;
+    if (!range->right->is<IR::Constant>()) {
+        ::error("%1%: must evaluate to a compile-time constant.",
+                range->right);
+        return false;
+    }
+    auto right = range->right->to<IR::Constant>()->value;
+    if (right < left) {
+        if (base == 16) {
+            std::stringstream bufferL;
+            bufferL << std::hex << std::showbase << left;
+            std::stringstream bufferR;
+            bufferR << std::hex << std::showbase << right;
+            ::error(ErrorType::ERR_INVALID, "%2% -%3%: Range end is less than start.",
+                    range, bufferL.str(), bufferR.str());
+        } else {
+            ::error(ErrorType::ERR_INVALID, "%2% -%3%: Range end is less than start.",
+                    range, left, right);
+        }
+        return false;
+    }
+    return true;
+}
+
+std::vector<const IR::Mask *>
+DoReplaceSelectRange::rangeToMasks(const IR::Range *r) {
+    bool st = checkRange(r);
+    if (!st)
+        ::error("Range check failed");
+
+    int width = typeMap->getType(r, true)->width_bits();
+    big_int min = r->left->to<IR::Constant>()->value;
+    big_int max = r->right->to<IR::Constant>()->value;
+    std::vector<const IR::Mask *> masks;
+    big_int range_size_remaining = max - min + 1;
+
+    while (range_size_remaining > 0) {
+        big_int range_size = ((big_int) 1) << trailingZeros(min, width);
+
+        while (range_size > range_size_remaining)
+            range_size >>= 1;
+
+        auto constType = typeMap->getType(r->left, true);
```

Comment:
I don't think you need the type map, `r->left->type` should have all the information you need.